File: wear/src/main/java/com/pdyjak/powerampwear/music_browser/artists/ArtistsBrowserFragment.java
package com.pdyjak.powerampwear.music_browser.artists;

import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.gms.wearable.MessageEvent;
import com.pdyjak.powerampwear.MessageExchangeHelper;
import com.pdyjak.powerampwear.MessageListener;
import com.pdyjak.powerampwear.R;
import com.pdyjak.powerampwear.music_browser.BrowserFragmentBase;
import com.pdyjak.powerampwear.music_browser.Clickable;
import com.pdyjak.powerampwear.music_browser.ItemViewHolder;
import com.pdyjak.powerampwear.music_browser.MusicLibraryNavigator;
import com.pdyjak.powerampwear.music_browser.ViewHolderFactory;
import com.pdyjak.powerampwearcommon.requests.RequestsPaths;
import com.pdyjak.powerampwearcommon.responses.Artist;
import com.pdyjak.powerampwearcommon.responses.ArtistsResponse;

import java.util.ArrayList;
import java.util.List;

public class ArtistsBrowserFragment extends BrowserFragmentBase implements MessageListener {

    private static class ArtistsViewHolderFactory implements ViewHolderFactory {
        @NonNull
        @Override
        public ItemViewHolder createViewHolder(ViewGroup parent) {
            View view = LayoutInflater.from(parent.getContext()).inflate(
                    R.layout.basic_two_line_item, parent, false
            );
            return new ArtistViewHolder(view);
        }
    }

    @NonNull
    @Override
    protected ViewHolderFactory createViewHolderFactory() {
        return new ArtistsViewHolderFactory();
    }

    @Override
    protected boolean tryRestoreCachedItems() {
        ArtistsResponse cached = getMusicLibraryCache().getArtists();
        if (cached != null) {
            setItems(transform(cached.getArtists()));
            return true;
        }
        return false;
    }

    @Override
    protected void fetchItems() {
        MessageExchangeHelper helper = getMessageExchangeHelper();
        helper.addMessageListenerWeakly(this);
        helper.sendRequest(RequestsPaths.GET_ARTISTS);
    }

    @Override
    public void onPause() {
        super.onPause();
        getMessageExchangeHelper().removeMessageListener(this);
    }

    @Override
    public void onMessageReceived(@NonNull MessageEvent messageEvent) {
        if (ArtistsResponse.PATH.equals(messageEvent.getPath())) {
            getMessageExchangeHelper().removeMessageListener(this);
            byte[] bytes = messageEvent.getData();
            if (bytes == null) return;
            ArtistsResponse response = ArtistsResponse.fromBytes(bytes);
            if (response != null) {
                setItems(transform(response.getArtists()));
                getMusicLibraryCache().update(response);
            }
        }
    }

    private List<Clickable> transform(@NonNull List<Artist> artists) {
        MusicLibraryNavigator navigator = getMusicLibraryNavigator();
        List<Clickable> items = new ArrayList<>();
        for (Artist artist : artists) {
            items.add(new ArtistItem(navigator, artist.id, artist.name, artist.songsCount));
        }
        return items;
    }
}

File: wear/src/main/java/com/pdyjak/powerampwear/music_browser/artists/ArtistsBrowserFragment.kt
package com.pdyjak.powerampwear.music_browser.artists

import android.view.LayoutInflater
import android.view.ViewGroup

import com.google.android.gms.wearable.MessageEvent
import com.pdyjak.powerampwear.MessageListener
import com.pdyjak.powerampwear.R
import com.pdyjak.powerampwear.messageExchangeHelper
import com.pdyjak.powerampwear.musicLibraryCache
import com.pdyjak.powerampwear.musicLibraryNavigator
import com.pdyjak.powerampwear.music_browser.BrowserFragmentBase
import com.pdyjak.powerampwear.music_browser.ItemViewHolder
import com.pdyjak.powerampwear.music_browser.ViewHolderFactory
import com.pdyjak.powerampwearcommon.requests.RequestsPaths
import com.pdyjak.powerampwearcommon.responses.Artist
import com.pdyjak.powerampwearcommon.responses.ArtistsResponse

class ArtistsBrowserFragment : BrowserFragmentBase(), MessageListener {

    private class ArtistsViewHolderFactory : ViewHolderFactory {
        override fun createViewHolder(parent: ViewGroup): ItemViewHolder {
            val view = LayoutInflater.from(parent.context).inflate(
                    R.layout.basic_two_line_item, parent, false
            )
            return ArtistViewHolder(view)
        }
    }

    override fun createViewHolderFactory(): ViewHolderFactory {
        return ArtistsViewHolderFactory()
    }

    override fun tryRestoreCachedItems(): Boolean {
        val cached = activity.musicLibraryCache.artists
        if (cached !== null) {
            setItems(transform(cached.getArtists()))
            return true
        }
        return false
    }

    override fun fetchItems() {
        val helper = activity.messageExchangeHelper
        helper.addMessageListenerWeakly(this)
        helper.sendRequest(RequestsPaths.GET_ARTISTS)
    }

    override fun onPause() {
        super.onPause()
        activity.messageExchangeHelper.removeMessageListener(this)
    }

    override fun onMessageReceived(messageEvent: MessageEvent) {
        if (ArtistsResponse.PATH == messageEvent.path) {
            activity.messageExchangeHelper.removeMessageListener(this)
            val bytes = messageEvent.data ?: return
            val response = ArtistsResponse.fromBytes(bytes)
            if (response !== null) {
                setItems(transform(response.artists))
                activity.musicLibraryCache.update(response)
            }
        }
    }

    private fun transform(artists: List<Artist>): List<ArtistItem> {
        val navigator = activity.musicLibraryNavigator
        return artists.map { ArtistItem(navigator, it.id, it.name, it.songsCount) }
    }
}
